### PR TITLE
blog: wip development vs harness development

### PR DESCRIPTION
**Pull request by nicks**
Hello @ellenkorbes,

Please review the following commits I made in branch nicks/wip:

0666039fc4e2bf9357f415e492056a966260ddc2 (2021-05-26 19:48:57 -0400)
blog: wip development vs harness development

Code review reminders, by giving a LGTM you attest that:

* Commits are adequately tested
* Code is easy to understand and conforms to style guides
* Incomplete code is marked with TODOs
* Code is suitably instrumented with logging and metrics

### CODE REVIEW

**Comment by nicks** [comment]
Goals:
- in the spirit of blog posts as documentation, I want to try the pattern of using slack discussions as a pipe into blog posts, so we have things to point to in the future
- I'm trying to test out different names of what Tilt is for, and see what terms people repeat back to me. I really admire "GitOps" as a term of art. I like how "WIP changes" (a) mirrors the "WIP" commit message everyone uses, (b) clarifies that you really shouldn't deploy with Tilt, and (c) sounds funny

File: blog/_posts/2021-05-26-wip-vs-harness.md [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+image: "/assets/images/wip-vs-harness/cover.jpg"
+image_caption: "<i>Everybody Rides in the Whip at Midland Beach</i>. Via <a href='https://digitalcollections.nypl.org/items/510d47d9-c024-a3d9-e040-e00a18064a99'>The New York Public Library</a>."
+tags:
+  - gitops
+  - localdev
+---
+
+In the #tilt chat last week, `renaudguerin` asked:
+
+> I’m the new DevOps guy in a small startup, and migrating us away from broken
```
Comment (ellenkorbes) [inline blog/_posts/2021-05-26-wip-vs-harness.md:18]: I would replace "away from broken in-house tooling" with an ellipsis, tbqh. Don't wanna be broadcasting someone criticize their own company.

File: blog/_posts/2021-05-26-wip-vs-harness.md [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+
+- How these two toolsets approach those problems
+
+- How to bridge the gap between them
+
+## More Service, More Problems
+
+If you're working on a multi-service app, you have two types of changes:
+
+- WIP changes - No services are running in dev, and you want to make a change.
```
Comment (ellenkorbes) [inline blog/_posts/2021-05-26-wip-vs-harness.md:53]: I understand what the two types are from looking at the table below, but these two sentences of explanation are *very* confusing to me. Maybe I'm missing something, or maybe there's an unlucky typo somewhere? I don't know.

File: blog/_posts/2021-05-26-wip-vs-harness.md [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+These two sets may sound the same but they are mostly *non-overlapping*!
+
+Here's a table to help illustrate the differences:
+
+|          | WIP Changes | Harness Changes |
+----------------|-------------|------------------
+| Add segfaults to see what breaks                 | ✅ | ❌️                             |
+| Bring up all services from scratch               | ✅ | Something terrible has happened |
+| Reset the database                               | ✅️ | ❌ ❌ ❌                      |
+| Sacifice correctness for faster updates | ✅  |     ❌️                            |
```
Comment (ellenkorbes) [inline blog/_posts/2021-05-26-wip-vs-harness.md:66]:
```suggestion
| Sacrifice correctness for faster updates | ✅  |     ❌️                            |
```

File: blog/_posts/2021-05-26-wip-vs-harness.md [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+| Add security backdoors for testing | ✅ | 🚑 |
```
Comment (ellenkorbes) [inline blog/_posts/2021-05-26-wip-vs-harness.md:67]: Can we rephrase this to "security is not a concern" or something more neutral like that? "Add security backdoors for testing" has a very let's-kick-some-puppies feel to me.

File: blog/_posts/2021-05-26-wip-vs-harness.md [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+
+- When your file system changes, the Tilt server updates your dev env to match
+  what's on disk.
+
+- Tilt doesn't worry to much about recording the history of changes, assuming
+  that you can throw it all away when you're done.
+
+## Blurred Lines
+
+We recommend you use both GitOps and local K8s together! Tilt-team does this.
```
Comment (ellenkorbes) [inline blog/_posts/2021-05-26-wip-vs-harness.md:96]:
```suggestion
We recommend you use both GitOps and local K8s together! The Tilt team does this.
```

File: blog/_posts/2021-05-26-wip-vs-harness.md [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+We even work with teams who use Tilt to run ArgoCD in dev, to make changes to
+how their GitOps pipeline works together.
+
+But there are teams who mix and match these tools in more creative ways!
+
+---
+
+### Local K8s for Harness Development
+
+- I once met a team who needed to deploy to an on-prem mainframe. They created
```
Comment (ellenkorbes) [inline blog/_posts/2021-05-26-wip-vs-harness.md:113]: What is this paragraph getting at?

**Comment by ellenkorbes** [comment]
I think overall what's missing is a clear and concise statement of the point of the two.

After reading I get the feeling that Harness-type is good and reliable vs. WIP-type which is sloppy and insecure. That's because the end-goal of each wasn't outlined explicitly, only their characteristics. And I don't we can expect everyone to connect the dots themselves.

So for example: 

* "WIP-style serves for fast iteration, when you're actively working on a solution to a problem, and you want as-fast-as-possible feedback, and it's a development environment anyway so correctness and reliability are secondary concerns to sheer speed of iteration."
* "Harness-style is what happens after the step above. It's when the solutions to individual coding steps have already been solved, and now we're trying to conform that code to the standards of a larger system. Here we are very concerned about correctness, reliability, and security, but we don't care about speed anymore."

And I would also point out that the point in which these two connect is the commit/PR. One workflow is focused on the pre-commit stage, the other on the post-commit stage.

**Comment by nicks** [comment]
ya, I think that's a fair high-level criticism. Let me think on this a bit.

File: blog/_posts/2021-05-26-wip-vs-harness.md (hunk 6 above)
Comment (nicks) [reply blog/_posts/2021-05-26-wip-vs-harness.md:113]: there ARE teams who use "tools for WIP/pre-commit dev" to deploy their production stuff. it's not...wrong, just a weird use of that tooling.

In the early NodeJS days, there were major problems where people would serve the `.git` directories of their code in production, and accidentally leak their whole git history :grimacing:

File: blog/_posts/2021-05-26-wip-vs-harness.md (hunk 1 above)
Comment (nicks) [reply blog/_posts/2021-05-26-wip-vs-harness.md:18]: done

File: blog/_posts/2021-05-26-wip-vs-harness.md (hunk 6 above)
Comment (nicks) [reply blog/_posts/2021-05-26-wip-vs-harness.md:122]: i added some context above

**Comment by nicks** [comment]
ahhh, i keep getting sidelined when i set aside time to work on this, so just made the edits you suggested!

**Review by ellenkorbes — APPROVED** [review]
LGTM
